src/object_store/factory.rs

```diff
@@ -0,0 +1,218 @@
+) -> Result<Arc<dyn ObjectStore>, object_store::Error> {
+    Ok(match &object_store_cfg {
+        schema::ObjectStore::Local(schema::Local { data_dir }) => {
+            Arc::new(LocalFileSystem::new_with_prefix(data_dir)?)
+        }
+        schema::ObjectStore::InMemory(_) => Arc::new(InMemory::new()),
+        #[cfg(feature = "object-store-s3")]
+        schema::ObjectStore::S3(S3 {
+            region,
+            access_key_id,
+            secret_access_key,
+            session_token,
+            endpoint,
+            bucket,
+            ..
+        }) => {
+            let mut builder = AmazonS3Builder::new()
+                .with_region(region.clone().unwrap_or_default())
+                .with_bucket_name(bucket)
+                .with_allow_http(true);
+
+            if let Some(endpoint) = endpoint {
+                builder = builder.with_endpoint(endpoint);
+            }
+
+            if let (Some(access_key_id), Some(secret_access_key)) =
+                (&access_key_id, &secret_access_key)
+            {
+                builder = builder
+                    .with_access_key_id(access_key_id)
+                    .with_secret_access_key(secret_access_key);
+
+                if let Some(token) = session_token {
+                    builder = builder.with_token(token)
+                }
+            } else {
+                builder = builder.with_skip_signature(true)
+            }
+
+            let store = builder.build()?;
+
+            if let Some(props) = cache_properties {
+                props.wrap_store(Arc::new(store))
+            } else {
+                Arc::new(store)
+            }
+        }
+        #[cfg(feature = "object-store-gcs")]
+        schema::ObjectStore::GCS(GCS {
+            bucket,
+            google_application_credentials,
+            ..
+        }) => {
+            let gcs_builder: GoogleCloudStorageBuilder =
+                GoogleCloudStorageBuilder::new().with_bucket_name(bucket);
+
+            let gcs_builder = if let Some(path) = google_application_credentials {
+                gcs_builder.with_service_account_path(path)
+            } else {
+                gcs_builder
+            };
+
+            let store = gcs_builder.build()?;
+
+            if let Some(props) = cache_properties {
+                props.wrap_store(Arc::new(store))
+            } else {
+                Arc::new(store)
+            }
+        }
+    })
+}
+
+#[derive(PartialEq, Eq)]
+struct StoreCacheKey {
+    url: Url,
+    options: HashMap<String, String>,
+}
+
+impl Hash for StoreCacheKey {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        Hash::hash(&self.url, state);
+
+        let mut pairs: Vec<_> = self.options.iter().collect();
+        pairs.sort_by_key(|i| i.0);
+
+        Hash::hash(&pairs, state);
+    }
+}
+
+pub struct ObjectStoreFactory {
+    default_store: Arc<InternalObjectStore>,
+    custom_stores: RwLock<HashMap<StoreCacheKey, Arc<dyn ObjectStore>>>,
```

Comment:
Let's consider doing this, though we can do it in a follow-up PR.
```suggestion
    custom_stores: DashMap<StoreCacheKey, Arc<dyn ObjectStore>>>,
```